### PR TITLE
edk2: 202402 -> 202405

### PR DESCRIPTION
**Pull request by adamcstephens**
## Description of changes

https://github.com/tianocore/edk2/releases/tag/edk2-stable202405

- **edk2: simplify src by using postFetch**
- **edk2: move to by-name**
- **edk2: add updateScript and changelog**

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by adamcstephens** [comment]
Result of `nixpkgs-review pr 321143` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>OVMF</li>
    <li>OVMF-cloud-hypervisor</li>
    <li>OVMF-cloud-hypervisor.fd</li>
    <li>OVMF.fd</li>
    <li>OVMFFull</li>
    <li>OVMFFull.fd</li>
    <li>edk2</li>
    <li>edk2-uefi-shell</li>
    <li>lxd-lts</li>
    <li>multipass</li>
    <li>quickemu</li>
    <li>quickgui</li>
  </ul>
</details>

File: pkgs/by-name/ed/edk2/package.nix [hunk 1]
```diff
@@ -48,29 +46,27 @@ edk2 = stdenv.mkDerivation rec {
...
+    # We don't want EDK2 to keep track of OpenSSL,
+    # they're frankly bad at it.
+    postFetch = ''
+      rm -rf $out/CryptoPkg/Library/OpensslLib/openssl
+      mkdir -p $out/CryptoPkg/Library/OpensslLib/openssl
+      tar --strip-components=1 -xf ${buildPackages.openssl.src} -C $out/CryptoPkg/Library/OpensslLib/openssl
+
+      # Fix missing INT64_MAX include that edk2 explicitly does not provide
+      # via it's own <stdint.h>. Let's pull in openssl's definition instead:
+      sed -i $out/CryptoPkg/Library/OpensslLib/openssl/crypto/property/property_parse.c \
```
Comment (Mic92) [inline pkgs/by-name/ed/edk2/package.nix:65]: This maybe look a bit simpler, but don't we have now to re-fetch everytime, we want to patch anything? An alternative would be to move modifications to patchPhase instead.

Comment (Mic92) [reply pkgs/by-name/ed/edk2/package.nix]: Thoughts on https://github.com/NixOS/nixpkgs/pull/321143/commits/8cd9620f0d7bfc76ed4707978b26af0d6889ce16 ?

Comment (adamcstephens) [reply pkgs/by-name/ed/edk2/package.nix]: I think OpenSSL needs to be replaced in src since the src derivation is reused below in edk2.mkDerivation

Comment (adamcstephens) [reply pkgs/by-name/ed/edk2/package.nix]: I can do some more testing. My goal was to fix the automatic updating

Comment (adamcstephens) [reply pkgs/by-name/ed/edk2/package.nix]: Yeah, `nix-update` doesn't know how to handle the source hash not being in `src`. 

```
nix_update.errors.UpdateError: Could not find a url in the derivations src attribute
```

I think our options are either:

1. Move back to the separate `runCommand` wrapper we had before this PR, and write a custom update script
2. Accept that we'll have to re-fetch the source if we want to modify the content in that previous `runCommand` (e.g. the openssl replacement I put in `postFetch` and you moved to `postPatch`)

Comment (RaitoBezarius) [reply pkgs/by-name/ed/edk2/package.nix]: That doesn't work @Mic92, OVMF will use edk2.src, not the edk2.src post patched.

Comment (adamcstephens) [reply pkgs/by-name/ed/edk2/package.nix:65]: I've removed your commit @Mic92 as it won't work as expected.

**Comment by RaitoBezarius** [comment]
```
error: hash mismatch in fixed-output derivation '/nix/store/lmr9v0wlgrm41zb9cgwvqlj7mg43b7zw-source.drv':
         specified: sha256-7vNodHocwqQiO0ZXtqo8lEOFyt8JkFHcAathEhrKWE0=
            got:    sha256-4JDeA4QBa/FYqgHi5l6FT9EdcTtFvIbKzO+ywT1BsC8=
```

OVMF does not build anymore.

File: pkgs/by-name/ed/edk2/package.nix [hunk 2]
```diff
@@ -48,29 +46,27 @@ edk2 = stdenv.mkDerivation rec {
...
   ];
 
-  srcWithVendoring = fetchFromGitHub {
+  src = fetchFromGitHub {
     owner = "tianocore";
     repo = "edk2";
     rev = "edk2-stable${edk2.version}";
     fetchSubmodules = true;
-    hash = "sha256-Nurm6QNKCyV6wvbj0ELdYAL7mbZ0yg/tTwnEJ+N18ng=";
+    hash = "sha256-7vNodHocwqQiO0ZXtqo8lEOFyt8JkFHcAathEhrKWE0=";
```
Comment (superherointj) [inline pkgs/by-name/ed/edk2/package.nix:54]: Please verify if hash here should be:
`sha256-4JDeA4QBa/FYqgHi5l6FT9EdcTtFvIbKzO+ywT1BsC8=`

Comment (superherointj) [reply pkgs/by-name/ed/edk2/package.nix:54]: Would it make sense to add `edk2` package to `openssl` passthru.tests, such that, whenever someone is bumping `openssl` is also going to be hinted that `edk2` hash should be bumped too?

**Comment by Mic92** [comment]
Fix is here: https://github.com/NixOS/nixpkgs/pull/325267

**Pull request merged**